### PR TITLE
AIRFLOW-1551 : Add an operator to trigger job in jenkins from Airflow

### PR DESCRIPTION
**Pull request by moe-nadal-ck**
Add a new operator to trigger job on jenkins.
A new connection type has also been added with it's companion hook.

No test are written yet as I don't know how to write test in Python, but definitely interested by example or how to to add some. 

Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Airflow JIRA 1551 ](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1551

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
Add a new connection type to configure jenkins connection. This type of connection takes the following parameters : 
- host
- login
- api token/password (both can be used)
- port

This connection is associated to a hook to build the Jenkins object (from python-jenkins library)

An operator allows to trigger job on the remote jenkins. It accept parameters as a python dict or as a file (which should contains the python dict). The parameters are declared as template field to allow usage of macros in them. 

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
One test written for the operator, but it doesn't seem triggered and I don't know why. 

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by moe-nadal-ck** [comment]
This PR still needs test, but I don't know how to write python test/test for airflow. Do you have any example on how to write connection/hook/operators tests ?
And should I be using a public demo jenkins instance or should I mock it ? (the latter would require way more work).

**Comment by Acehaidrey** [comment]
https://github.com/apache/incubator-airflow/tree/master/tests/operators
Have you looked at these tests? They mock connections so I think you may have to look into that too

**Comment by moe-nadal-ck** [comment]
So I test is written to test the operator, I'm definitely welcoming any comment or idea to improve either the test or the operator :). 

I haven't been able to run the unit test locally, but I'm hoping Travis will be running them. Is there any way to check if my new test has been running ? (after checking the log, I don't see anything about my new test, did I do something wrong ? )

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (jgao54) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: nit: it may be better to pass in an arg like max_retry_attempt or max_wait_time instead of hard-coding retry attempts = 10, so that the total retry time is a bit more flexible and doesn't have to depend on sleep_time.

Comment (moe-nadal-ck) [reply airflow/contrib/operators/jenkins_job_trigger_operator.py]: Thanks, that's fixed :)

**Comment by jgao54** [comment]
Looks like tests are failing in python3 envs (basestring no longer exists in python3):
```10) ERROR: test_execute (tests.JenkinsOperatorTestCase)
----------------------------------------------------------------------
   Traceback (most recent call last):
    tests/contrib/operators/test_jenkins_operator.py line 67 in test_execute
      operator.execute(None)
    airflow/contrib/operators/jenkins_job_trigger_operator.py line 74 in execute
      if isinstance(self.parameters, basestring):
   NameError: name 'basestring' is not defined
   -------------------- >> begin captured logging << --------------------
   root: INFO: Triggering the job a_job_on_jenkins on the jenkins : fake_jenkins_connection with the parameters : {'another_param': '42', 'a_param': 'blip'}
   --------------------- >> end captured logging << ---------------------```

**Comment by moe-nadal-ck** [comment]
Yep, I just fixed that too with the last commit. (it's the first time I wrote some python, the tests have been pretty painful to make them working xD, but I think I get closer :) )

File: airflow/contrib/hooks/jenkins_hook.py
Comment (skudriashev) [inline airflow/contrib/hooks/jenkins_hook.py]: Use `"""` instead of `'''`

File: airflow/contrib/hooks/jenkins_hook.py
Comment (skudriashev) [inline airflow/contrib/hooks/jenkins_hook.py]: Remove print statement.

File: airflow/contrib/hooks/jenkins_hook.py
Comment (skudriashev) [inline airflow/contrib/hooks/jenkins_hook.py]: Remove print.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: This line is to long, please try to fit in 80 symbols.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Parenthesis are redundant.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Parenthesis are redundant.

**Comment by moe-nadal-ck** [comment]
The tests are passing 🎉  🎉  🎉  🎉  🎉  🎉  🎉  🎉  🎉  🎉  🎉  🎉  
🎆   😄  😸  🎉  🎉  🎉  🎉  🎉  🎉  🎉  🎉

File: airflow/models.py
Comment (skudriashev) [inline airflow/models.py]: It is better to insert in alphabetical order.

File: airflow/contrib/hooks/__init__.py
Comment (skudriashev) [inline airflow/contrib/hooks/__init__.py]: Please keep ',' at the end of line.

**Comment by jgao54** [comment]
You may want to squash the commits :P

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: 2 empty lines before class definition (pep8)

**Comment by moe-nadal-ck** [comment]
:P yep, I was just waiting for the test to pass before the final cosmetic pass :)

**Review by skudriashev — CHANGES_REQUESTED** [review]

File: airflow/contrib/hooks/jenkins_hook.py
Comment (moe-nadal-ck) [reply airflow/contrib/hooks/jenkins_hook.py]: Should I transform this last one in a logging.info/debug ? It can be useful to have the full url used in case of error when configuring the connection.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (moe-nadal-ck) [reply airflow/contrib/operators/jenkins_job_trigger_operator.py]: I tried to cut all the longest lines of the file. 
Is the proper way to do it is to add \ at the end of the line before the line return ?

File: airflow/contrib/hooks/jenkins_hook.py
Comment (skudriashev) [reply airflow/contrib/hooks/jenkins_hook.py]: In this case yes please convert to logging

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Is empty name acceptable?

Comment (moe-nadal-ck) [reply airflow/contrib/operators/jenkins_job_trigger_operator.py]: I raise an exception after is the name is empty. Should I remove the default value to force the user to give a name ?

Comment (moe-nadal-ck) [reply airflow/contrib/operators/jenkins_job_trigger_operator.py]: Default argument removed, I hope the tests will still pass :D

**Comment by moe-nadal-ck** [comment]
I just removed the empty default argument to force the user to specify them and I think I processed all the comments. 
Tests are passing correctly, I think I'm ready to get the PR merge on my side :).

File: airflow/contrib/hooks/jenkins_hook.py
Comment (skudriashev) [inline airflow/contrib/hooks/jenkins_hook.py]: Please add possibility to use https.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Backslash is redundant.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Backslash is redundant.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Backslash is redundant.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Backslash is redundant.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (skudriashev) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Please use string formatting instead of concatenation.

**Review by skudriashev — CHANGES_REQUESTED** [review]
Please fit all lines at 80 symbols (pep8) or in worst case 100 (max 120) symbols to increase readability. Thanks.

File: airflow/contrib/hooks/jenkins_hook.py
Comment (moe-nadal-ck) [reply airflow/contrib/hooks/jenkins_hook.py]: Done, I used the field extra of the connection to store a boolean for that.

**Comment by moe-nadal-ck** [comment]
@skudriashev Hey, thanks for all these comments :), for the line size, does the 80 characters limit include the indentation or not ? 
And when I add a line return in the middle of a call, should I align the new line exactly or add an indentation to show it's the end of the previous line ? 
(as you can imagine, I'm pretty new to python... :) )

**Comment by moe-nadal-ck** [comment]
I had to add back the \ as if I don't put them in my strings, I have syntax error. I could use multiline string instead but I don't want line break in the middle of error messages :P.

**Comment by skudriashev** [comment]
@moe-nadal-ck how could you have syntaxt error? Please see other hook and how long lines are handled.

**Comment by moe-nadal-ck** [comment]
Fixed, I wasn't closing/reopening my string (I just had a line return in the middle of a one line string). It's fixed now :)

**Comment by moe-nadal-ck** [comment]
Just added an example on how to use the operator and how to retrieve artifact (I hesitate to put it directly in the code of the operator, but I don't think that many people would need that...). 
Also made a small change to store the job url in an xcom. 
And normally all comment have been processed.

**Comment by moe-nadal-ck** [comment]
Rebase on main branch to update the logging logic (updated by [AIRFLOW-1582] Improve logging within Airflow in master)

File: airflow/contrib/hooks/jenkins_hook.py
Comment (skudriashev) [inline airflow/contrib/hooks/jenkins_hook.py]: What do you do here?

File: airflow/contrib/hooks/jenkins_hook.py
Comment (skudriashev) [inline airflow/contrib/hooks/jenkins_hook.py]: Use lazy formatting.

File: airflow/contrib/hooks/jenkins_hook.py
Comment (moe-nadal-ck) [reply airflow/contrib/hooks/jenkins_hook.py]: I had a comment, but I was setting a default value in connection.extra to avoid having a ValueError thrown when I do the strtobool. 
The connection.extra field host the value to tell if we should use http or https (the label has been renamed in the connection form to something understandable).

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (bolkedebruin) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Add the mixin to the class please. Python supports multiple inheritance

Comment (moe-nadal-ck) [reply airflow/contrib/operators/jenkins_job_trigger_operator.py]: Done

Comment (bolkedebruin) [reply airflow/contrib/operators/jenkins_job_trigger_operator.py]: Sorry baseoperator already uses the mixin. You can remove it from the definition. (My mistake)

Comment (moe-nadal-ck) [reply airflow/contrib/operators/jenkins_job_trigger_operator.py]: Fixed, I'll pushed that as soon as I find how to do lazy logging in Python ^^ (it was definitely not a long thing to fix...^^ )

File: airflow/contrib/hooks/jenkins_hook.py
Comment (moe-nadal-ck) [reply airflow/contrib/hooks/jenkins_hook.py]: Done

**Comment by larrywongl** [comment]
Exactly what we want, but when can this merged into the base branch?

And we want to display the link to the Jenkins jobs and also render the Jenkins logs in airflow logs, can we do that

**Comment by moe-nadal-ck** [comment]
@larrywongl The link for the job is already in the airflow logs. I haven't put the full jenkins log as sometime they get pretty big (some of our jobs have multiple Gb of logs ^^). But I think having the link for the jenkins job is usually enough for navigability. 

And to get it merged with master branch, I've adressed all the comments I've receive, but still open to proposition. If that looks good to everybody, we can merge it I think :).

File: airflow/contrib/hooks/jenkins_hook.py
Comment (love3) [inline airflow/contrib/hooks/jenkins_hook.py]: seems a typo here. should be connection.password not connection._password

**Comment by larrywongl** [comment]
@moe-nadal-ck, so are u able to put a hyper link of the Jenkins jobs in the logs? I tried several times but none of them are work. let me know if you could

Could you please push for the merge effort, or we have to implement this by ourselves which is quite inefficient.

File: airflow/contrib/hooks/jenkins_hook.py
Comment (moe-nadal-ck) [reply airflow/contrib/hooks/jenkins_hook.py]: Fixed, both are working, but password seems more frequent in the code base so I switch to this one :). Good catch ;)

**Comment by moe-nadal-ck** [comment]
@larrywongl : I'm doing my best to get this merged, but it's my first contribution to this project (and also the first time I'm writing python) so I prefer to has as much eyes on it as possible to ensure a good quality  :). 
For the url, I've been using the python-jenkins library to trigger the job on jenkins, and the URL for the job is exposed through the build_info object : 
`build_info = jenkins_server.get_build_info(job_name=self.job_name, build_number=build_number)`
`build_info['url]' # <- contains the URL of the job, add /log at the end for a direct link to the logs`

And if you need the plugin before it get merged, you can still pick the commit (or only the file it contains and inject that in your airflow installation, that what we're doing at CK until a new version of airflow contains the operator :) )

**Comment by moe-nadal-ck** [comment]
@skudriashev @bolkedebruin @jgao54 Do you have any other comment/review on things I can still enhance before we can merge this PR ? I would like to be able to make it for the 1.8.3 (if possible :D ).
And thanks again for all your comment and your help for this PR ;)

**Comment by moe-nadal-ck** [comment]
The Travis hook seems broken, all the last builds passed but github displays a failure. Do anyone know how to fix that ?

**Comment by skudriashev** [comment]
@moe-nadal-ck you don't have your hook defined:

```
NameError: global name 'JenkinsHook' is not defined
```

**Comment by moe-nadal-ck** [comment]
That was an import problem in a old commit. For the 5 last builds, it was ok, but git hub stay stuck on a really old one (the last one is the 60 and all tests pass perfectly : https://travis-ci.org/moe-nadal-ck/incubator-airflow ). 
I'll try to close/re-open the PR once to see if it fixes the problem.

**Comment by moe-nadal-ck** [comment]
Finally found the problem, an import of LoggingMixing was refused when it was run from the master branch (run from my branch only were working, I don't know exactly why). I removed all the import for LoggingMixin as the logger is available in BaseOperator and now everything is fine :).

**Comment by Fokko** [comment]
@moe-nadal-ck Thanks for reporting. While writing Airflow-1611 I found that it is possible to end up with circular dependency imports. Also the `SalesforceHook` does not require an explicit import from the LoggingMixin, since this one is already available by the BaseHook. I'll look into this as part of ticket Airflow-1605.

**Comment by moe-nadal-ck** [comment]
Are we ready to merge this PR ? and if yes, can someone do it ? :D

**Comment by larrywongl** [comment]
Are we able to stop the Jenkins from airflow via this operator at will, say for example, we start the DAG and for the moment I want to cancel a job which is already in running? and can we kick out completed jobs individually manually?

**Comment by moe-nadal-ck** [comment]
@larrywongl We can't cancel a job for now, not sure if the API expose the feature (but it might). But for what I've seen, it would be way more complex to code (I'm not sure if there's a hook for when a task failed because of timeout). But you can go manually in the jenkins interface and abort the job and that will make the airflow task to fail.

Edit : After a quick search https://stackoverflow.com/questions/21021905/how-to-stop-a-build-in-jenkins-via-the-rest-api to stop a job through API, I don't think it's already exposed in the library but it's possible to send the request at least. And this PR https://github.com/apache/incubator-airflow/pull/2644 does something similar. But I think it would be better to do this in another PR to merge this one as it is.

**Comment by larrywongl** [comment]
It seems we need this function as for now, we turned focus on airflow hence do not want to go back to work on Jenkins.

**Comment by moe-nadal-ck** [comment]
If you need a feature, create a ticket for it in the bug tracker of airflow and you're more than welcome to write the code for it, that's the principle of open source project. I think this PR cover pretty well the scope of the initial ticket : being able to trigger a job on a jenkins server from airflow. And as shown in the example dag I put in the PR, you can also use a basic python operator to take other actions than the one shipped with the operator.

File: airflow/contrib/operators/jenkins_job_trigger_operator.py
Comment (bolkedebruin) [inline airflow/contrib/operators/jenkins_job_trigger_operator.py]: Please use self.log. Self.logger is deprecated

Comment (moe-nadal-ck) [reply airflow/contrib/operators/jenkins_job_trigger_operator.py]: Done :)

**Comment by bolkedebruin** [comment]
Can you rebase? Then I will merge

**Comment by moe-nadal-ck** [comment]
@bolkedebruin I just rebase again to be sure to be up-to-date, can you merge ? :)

EDIT : (Actually do not merge, cf last comment)

**Comment by moe-nadal-ck** [comment]
Actually, wait, I'm testing it with the last changes I made after the PR comments and some of them are breaking the operator (some keyword argument are wrong). I will have to push some fix tomorrow.

**Comment by moe-nadal-ck** [comment]
@bolkedebruin It should be ok now, I fixed name argument, but I wonder why the test didn't catch that before. That's kind of strange.

EDIT: Ok, I got it, I was mocking the jenkins library in the test, and this way hiding the crash. Actually I was even making sure it will crash as the test were explicitely waiting for an argument with the wrong name. That's fixed now

File: airflow/contrib/hooks/salesforce_hook.py
Comment (Fokko) [inline airflow/contrib/hooks/salesforce_hook.py]: This `LoggingMixin` can be removed since you already inherit it from the BaseHook

Comment (moe-nadal-ck) [reply airflow/contrib/hooks/salesforce_hook.py]: I revert the line removing in Salesforce_hook as this PR is not related at all to anything in Salesforce. The change was induced by a conflict when I rebased on master a few weeks ago.

**Comment by moe-nadal-ck** [comment]
Just rebased another time to stay sync with master, and removed the change in salesforce_hook (line removing) as it was only noise.

**Comment by larrywongl** [comment]
How can we get context the same way like python operator? we want to get the task_id, execution_id when using the Jenkins operator

**Comment by moe-nadal-ck** [comment]
@larrywongl You can use jinja templates in the parameter block, and they give you access to all of that. As there is no python_callable to give, it's not really comparable to python operator.

**Comment by larrywongl** [comment]
Do you have detail steps on how to do that?

**Comment by larrywongl** [comment]
We hit another issue, if we are about to kick out same build concurrently at same time, the code will be out of sync as you have not lock the resources.

From: Wang, Larry
Sent: Friday, October 13, 2017 1:45 PM
To: 'apache/incubator-airflow' <reply@reply.github.com>; 'apache/incubator-airflow' <incubator-airflow@noreply.github.com>
Cc: 'Mention' <mention@noreply.github.com>; Wang, Ming <Ming.Wang@emc.com>
Subject: RE: [apache/incubator-airflow] AIRFLOW-1551 : Add an operator to trigger job in jenkins from Airflow (#2553)

Do you have detail steps on how to do that?

**Comment by moe-nadal-ck** [comment]
@bolkedebruin Can we merge the PR ? I'll take a look at the problem rised by larry in another PR if I can reproduce it, but I think it only appears when triggering a lot of job in parallel.

**Comment by moe-nadal-ck** [comment]
@bolkedebruin (or @skudriashev , @Fokko or @jgao54 , not sure who has or not merge credentials) I just rebase on master to resolve a conflict. Would it be possible to merge the PR ?

**Comment by moe-nadal-ck** [comment]
Important change in the PR to fix a bug coming from a misunderstanding of how jenkins attribute build number. The build number is given only once the job get executed and not when the job is getting triggered, so we need to poll the jenkins queue until the job execution start for that. 
This required to change slightly the way a request is sent to jenkins as the location in the queue is only present in the response headers. 
It is now fixed and I'm currently testing it. I hope @larrywongl will be able to test it in a state closer to airflow current HEAD (I'm using airflow 1.8.1 so the logging mechanism is slightly different, but normally I have take care of that :) )

**Comment by moe-nadal-ck** [comment]
Hi all (@jgao54 @Fokko @bolkedebruin @skudriashev )

I just rebase this branch and fixed all the flake8 issues to make it mergeable. Would it be possible to do it ? 
We've been using the operator at Credit Karma for a certain time now and no bug have been detected on our side. 

Thanks in advance, 
Moe

**Comment by Fokko** [comment]
Hi Moe, could you rebase your commits into one single commit with the `[AIRFLOW-1551]` prepended in the commit title? This is required to maintain history and keep everything nice and tidy. I'll test the PR locally. Cheers

**Comment by moe-nadal-ck** [comment]
run test

**Comment by unlikelyzero** [comment]
Any idea when this might be merged?

**Comment by Fokko** [comment]
@unlikelyzero yes, now! 

@moe-nadal-ck Thanks for your contribution!

**Pull request closed without merging**
